### ISSUE TITLE
tv3g3 PHP8 error with "Bulk update all titles"

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1401]
# BUG/ERROR report

## System information

* Tripal Version: 3.9dev
* Drupal Version: 7.94
* PostgreSQL Version: 13.3
* PHP Version: 8.0.27

### Issue description
What is happening is we are passing a job to a callback that does not expect it.
In file ```tripal/includes/TripalJob.inc``` function ```run()``` on this line
```
call_user_func_array($callback, $arguments);
```
the callback is ```tripal_update_all_urls_and_titles``` from file ```tripal/includes/tripal.bulk_update.inc```
which is ```function tripal_update_all_urls_and_titles($bundle_id, $update, $type) {```
and the arguments are
```
arguments="array(4) {
  ["bundle_id"]=>
  string(10) "bio_data_1"
  ["update"]=>
  string(21) "[local__abbreviation]"
  ["type"]=>
  string(5) "title"
  [0]=>
  object(TripalJob)#450 (8) {
    ["job_id":protected]=>
    NULL
    ["job":protected]=>
    object(stdClass)#451 (20) {
etc.
```
the fourth array element is ```$this```, appended at line 336 ```$arguments[] = $this;``` where, I think, we have checked for tripal v2 vs. v3, this is the v3 version.

#### Steps to reproduce
1. Need to run under PHP8.x
2. Go to organism content type /admin/structure/bio_data/manage/bio_data_1
3. Click the "Bulk update all titles" button
4. run the job with drush at the command line to see this output
```
2023-01-31 09:16:15: There are 1 jobs queued.
2023-01-31 09:16:15: Job ID 3953.
2023-01-31 09:16:15: Calling: tripal_update_all_urls_and_titles(bio_data_1, [local__abbreviation], title)
WD php: Error: Cannot use positional argument after named argument in TripalJob->run() (line 344 of          [error]
.../sites/all/modules/tripal/tripal/includes/TripalJob.inc).
Cannot modify header information - headers already sent by (output started at                                [warning]
.../sites/all/modules/tripal/tripal/api/tripal.jobs.api.inc:386)
bootstrap.inc:1551
Error: Cannot use positional argument after named argument in TripalJob->run() (line 344 of .../sites/all/modules/tripal/tripal/includes/TripalJob.inc).
Drush command terminated abnormally due to an unrecoverable error.                                           [error]
```

**Comment by dsenalik** [comment]
I don't fully understand all the possibilities that need to be handled. I might propose only appending ```$this``` when ```$arguments``` is a sequential array (but is this ever true?). This eliminates the error _in this specific case_
```
      if (count($refparams) > 0) {
        $lastparam = $refparams[count($refparams) - 1];
        if ($lastparam->getName() == 'job_id') {
          $arguments[] = $this->job->job_id;
        }
        else {
          // For tv3 append the job only if we are not using named arguments,
          // i.e. $arguments is a sequential array
          if (array_keys($arguments) == range(0, count($arguments)-1)) {
            $arguments[] = $this;
          }
        }
      }
```

**Comment by laceysanderson** [comment]
What about adding an optional argument of $job to the end of tripal_update_all_urls_and_titles()? I think that's what we typically did in Tripal 3 for all functions called by a job. It's passed in so the job can log + update percent.

**Comment by dsenalik** [comment]
I tried that
```
function tripal_update_all_urls_and_titles($bundle_id, $update, $type, $job) 
```
but because the other arguments are an associative array, and this one is appended without a "key", you still get this error
```
WD php: Error: Cannot use positional argument after named argument in TripalJob->run() (line 345 of    [error]
.../sites/all/modules/tripal/tripal/includes/TripalJob.inc).
```
You could use ```$arguments['job'] = $this;``` but I assume that would break everything else.

**Comment by laceysanderson** [comment]
Ahhhh, so I would do this to make the argument optional.

```
function tripal_update_all_urls_and_titles($bundle_id, $update, $type, $job = NULL) 
```

And yes, I would set the key. This shouldn't break anything as the arguments array is exploded before being passed to the calling function 🤔 
```
$arguments['job'] = $this;
```

**Comment by dsenalik** [comment]
For completeness, there is one more php8 error that I will fix
```
Trying to access array offset on value of type null tripal.bulk_update.inc:31
```

**Comment by dsenalik** [comment]
And finally, in pull #1402 I made this comment in the code actually be correct. Before, the message was printed for every item.
```
// Check if 50 items have been updated, if so print message.
```

**Comment by dsenalik** [comment]
Very much related, our blast jobs are failing, because the tripal_blast module is getting a "naked" job_id passed from the first part of this code
```
        if ($lastparam->getName() == 'job_id') {
          $arguments[] = $this->job->job_id;
        }
```
I thought that was just for tripal 2, but obviously some modules pass the job_id this way, so apparently we need to change this to
```
          $arguments['job_id'] = $this->job->job_id;
```

**Comment by spficklin** [comment]
I've not kept up with this conversation but I just merged PR #1402 .  If it's all fixed we can close but I'll let you make that call.

**Comment by dsenalik** [comment]
Known problems fixed with merged pull #1402
